File: Top Interview 150/Sliding Window/209_Minimum Size Subarray Sum.py
class Solution(object):
    def minSubArrayLen(self, target, nums):
        """
        :type target: int
        :type nums: List[int]
        :rtype: int
        """

        if sum(nums) < target:
            return 0
            
        curr_sum = 0
        min_len = len(nums) + 1

        start = 0
        end = 0

        while (end < len(nums)):
            while (curr_sum < target and end < len(nums)):
                curr_sum += nums[end]
                end += 1
            
            while (curr_sum >= target and start < len(nums)):
                if (end - start < min_len):
                    min_len = end - start
                
                curr_sum -= nums[start]
                start += 1

        return min_len

File: Top Interview 150/Sliding Window/209_Minimum Size Subarray Sum.cpp
class Solution {
public:
    int minSubArrayLen(int target, vector<int>& nums) {
        if (accumulate(nums.begin(), nums.end(), 0) < target) {
            return 0;
        }

        int curr_sum = 0;
        int n = nums.size();
        int min_len = n + 1;
        int start = 0;
        int end = 0;

        while (end < n) {
            while (curr_sum < target && end < n) {
                curr_sum += nums[end];
                end += 1;
            }
            while (curr_sum >= target && start < n) {
                if (end - start < min_len) {
                    min_len = end - start;
                }
                curr_sum -= nums[start];
                start += 1;
            }
        }
        return min_len;
    }
};